tests/components/sonos/test_services.py

```diff
@@ -1,53 +1,190 @@
 """Tests for Sonos services."""
 
+import asyncio
+from contextlib import asynccontextmanager
+import logging
 from unittest.mock import Mock, patch
 
 import pytest
 
-from homeassistant.components.media_player import DOMAIN as MP_DOMAIN, SERVICE_JOIN
+from homeassistant.components.media_player import (
+    DOMAIN as MP_DOMAIN,
+    SERVICE_JOIN,
+    SERVICE_UNJOIN,
+)
 from homeassistant.core import HomeAssistant
 from homeassistant.exceptions import HomeAssistantError
 
-from tests.common import MockConfigEntry
+from .conftest import MockSoCo, group_speakers, ungroup_speakers
 
 
 async def test_media_player_join(
-    hass: HomeAssistant, async_autosetup_sonos, config_entry: MockConfigEntry
+    hass: HomeAssistant,
+    sonos_setup_two_speakers: list[MockSoCo],
+    caplog: pytest.LogCaptureFixture,
 ) -> None:
-    """Test join service."""
-    valid_entity_id = "media_player.zone_a"
-    mocked_entity_id = "media_player.mocked"
+    """Test joining two speakers together."""
+    soco_living_room = sonos_setup_two_speakers[0]
+    soco_bedroom = sonos_setup_two_speakers[1]
+
+    # After dispatching the join to the speakers, the integration waits for the
+    # group to be updated before returning. To simulate this we will dispatch
+    # a ZGS event to group the speaker. This event is
+    # triggered by the firing of the join_complete_event in the join mock.
+    join_complete_event = asyncio.Event()
+
+    def mock_join(*args, **kwargs) -> None:
+        hass.loop.call_soon_threadsafe(join_complete_event.set)
+
+    soco_bedroom.join = Mock(side_effect=mock_join)
+
+    with caplog.at_level(logging.WARNING):
+        caplog.clear()
+        await hass.services.async_call(
+            MP_DOMAIN,
+            SERVICE_JOIN,
+            {
+                "entity_id": "media_player.living_room",
+                "group_members": ["media_player.bedroom"],
+            },
+            blocking=False,
+        )
+        await join_complete_event.wait()
+        # Fire the ZGS event to update the speaker grouping as the join method is waiting
+        # for the speakers to be regrouped.
+        group_speakers(soco_living_room, soco_bedroom)
+        await hass.async_block_till_done(wait_background_tasks=True)
```

Comment:
If the joining failed I think we should raise a `HomeAssistantError` instead so we can tell the user that without them having to look at the logs